### DISCUSSION TITLE
Full Stack Error Handling Type Safety (Either VS Errors)

### DISCUSSION THREAD

**Discussion by danielyogel** [trpc/trpc#1176] [General]
Was wondering, if anyone uses an [Either](https://gcanti.github.io/fp-ts/modules/Either.ts.html) type (together with always 200 status code) - instead of throwing - in order to achieve error handling type safety in the client?
Any reason why this may be a bad idea?

**Comment by KATT** [comment]
I've gone back-and-forth on this over the last few years. At the moment I prefer treating errors as exceptions.

You can return an `Either` but be aware that it'd also return a `200` response [and subsequently get cached if you have cache enabled] unless you do some custom stuff in the `responseMeta`.

Furthermore - you can do explicit error formatting in tRPC - not sure if you've played with [`formatError`](https://trpc.io/docs/error-formatting) yet.

**Reply by KATT** [reply]
Have a look at [the tests](https://github.com/trpc/trpc/blob/main/packages/server/test/errors.test.ts#L124-L295) for a deeper dive into `formatError`

**Reply by KATT** [reply]
Might be worth shooting the question out on the discord too - https://trpc.io/discord

**Reply by danielyogel** [reply]
Thanks!
The only client example I [see](https://trpc.io/docs/error-formatting), is using hooks in the client, but I use a vanilla client - so should I use something like 'waitError' (as used in the tests) to infer the error types in the client?

**Reply by KATT** [reply]
I see! The error formatting happens in `@trpc/client` so it's not an issue, but then I also understand why you want an `Either`.

You could do something like this https://github.com/trpc/trpc/commit/0379dbfa37b723a6e5a31af04ad29fd87db60091 - there's definitely ways of making this more generic so you could have a safe wrapper but I'll leave that to you to figure out :)

**Reply by KATT** [reply]
Actually - I took the demo a bit further, see here: https://github.com/trpc/trpc/pull/1188

**Reply by jedwards1211** [reply]
The shortcoming of `errorFormatter` is we don't get any typed error details specific to a procedure.  I often want this for deciding which form field(s) to show the error info under.  You can't always make form field names match your mutation input properties in complex forms.

I've been trying to think how to kludge something where I can *return* some error instance from a procedure so that its type gets passed to the client and add something to the serialization layer to transform it.  Then the question is if I could somehow still make that 4xx and throw an error on the client side while providing a utility function to extract the procedure-specific error type.

Or maybe I should just figure out something with `Either` and `responseMeta`, but I've definitely had to kludge together procedure-specific error type systems a lot throughout my years as a web developer

**Reply by jedwards1211** [reply]
There's no caching problem for mutations is there?  Getting a 200 on mutation failure is still undesirable though

**Reply by jedwards1211** [reply]
@KATT I figured out how to create a wrapper that accomplishes most of my goals here:
- Mutations reject if there was an error
- But, you can still get typed error information specific to the procedure from the error thrown

Basically I made a recursive proxy wrapper for the tRPC client that unwraps `Result` values returned from mutations.  If the mutation returns `{ success: false, error: ... }` it throws `new TypedProcedureError(result.error)`.  The recursive proxy provides an `isTypedError` type predicate next to `mutate`.  So you can use it like

```
  const trpc = useTRPC();
  const trpcClient = customizeTRPCClient(useTRPCClient());
  const navigate = useNavigate();

  const handleSubmit = createUserInvitationForm.useSubmit({
    onSubmit: async (values) => {
      try {
        const userInvitation =
          await trpcClient.userInvitation.create.mutate(values);
        await navigate({
          to: "/invitations",
          search: { userId: userInvitation.userId },
        });
      } catch (error) {
        if (trpcClient.userInvitation.create.isTypedError(error)) {
          if (error.err.code === "ALREADY_INVITED") {
            throw new z.ZodError([
              {
                path: createUserInvitationForm.get("email").path,
                code: z.ZodIssueCode.custom,
                message: "User has already been invited",
              },
            ]);
          }
        }
        throw error;
      }
    },
  });
``` 

More details here: https://gist.github.com/jedwards1211/ff8922db9ad124beb07940739482a49a

I would be really excited to see some kind of first-class typed procedure error data solution in tRPC.

**Comment by orendecor** [comment]
@danielyogel did you finally find a way to use Either with trpc? or you gave up

**Reply by danielyogel** [reply]
Hey @orendecor! I did, once back then, just returned an Either value. The problem is caching, I guess you'll have a problem to retry a failed operation, because the previous Either value would sometimes return from cache without hitting the server (haven't tried the cache problem my self).
